Pass IServiceProvider to configureBuilder of AddKafkaConsumer and AddKafkaProducer

g7ed6e opened the pull request:
Resolves #4323 

I had to deal with RS0027 and RS0028 and finally moved from optional parameters to explicit parameters overloads as parameter types differ.

sample usage:
```c#
var builder = Host.CreateApplicationBuilder(args);

builder.AddServiceDefaults();

builder.Services.AddSingleton<ISchemaRegistryClient>(provider =>
{
    var schemaRegistryConfig = new SchemaRegistryConfig
    {
        Url = "http://localhost:8081"
    };
    return new CachedSchemaRegistryClient(schemaRegistryConfig);
});
builder.AddKafkaProducer<Null, User>("kafka", configureBuilder: (provider, producerBuilder) =>
{
    var schemaRegistryClient = provider.GetRequiredService<ISchemaRegistryClient>();
    producerBuilder.SetValueSerializer(new AvroSerializer<User>(schemaRegistryClient));
});

builder.Services.AddHostedService<ContinuousProducerWorker>();

builder.Build().Run();
```



 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/4327)

g7ed6e committed:
Add overloads to AddKafkaConsumer and AddKafkaProducer extensions methods to use IServiceProvider in configureBuilder param

g7ed6e commented:
The public API modification should be binary backward compatible I followed guidance from https://github.com/dotnet/roslyn/blob/main/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md.

eerhardt commented:
Can you add some unit tests for all new API and functionality?

eerhardt commented:
>The public API modification should be binary backward compatible

@joperezr - thoughts on getting the package validation https://learn.microsoft.com/dotnet/fundamentals/apicompat/package-validation/overview hooked up in the repo now that we have shipped 8.0?

eerhardt commented on src/Components/Aspire.Confluent.Kafka/PublicAPI.Unshipped.txt:
```diff
@@ -19,7 +19,27 @@ Aspire.Confluent.Kafka.KafkaProducerSettings.DisableMetrics.get -> bool
-static Microsoft.Extensions.Hosting.AspireKafkaConsumerExtensions.AddKeyedKafkaConsumer<TKey, TValue>(this Microsoft.Extensions.Hosting.IHostApplicationBuilder! builder, string! name, System.Action<Aspire.Confluent.Kafka.KafkaConsumerSettings!>? configureSettings = null, System.Action<Confluent.Kafka.ConsumerBuilder<TKey, TValue>!>? configureBuilder = null) -> void
-static Microsoft.Extensions.Hosting.AspireKafkaProducerExtensions.AddKafkaProducer<TKey, TValue>(this Microsoft.Extensions.Hosting.IHostApplicationBuilder! builder, string! connectionName, System.Action<Aspire.Confluent.Kafka.KafkaProducerSettings!>? configureSettings = null, System.Action<Confluent.Kafka.ProducerBuilder<TKey, TValue>!>? configureBuilder = null) -> void
-static Microsoft.Extensions.Hosting.AspireKafkaProducerExtensions.AddKeyedKafkaProducer<TKey, TValue>(this Microsoft.Extensions.Hosting.IHostApplicationBuilder! builder, string! name, System.Action<Aspire.Confluent.Kafka.KafkaProducerSettings!>? configureSettings = null, System.Action<Confluent.Kafka.ProducerBuilder<TKey, TValue>!>? configureBuilder = null) -> void
+static Microsoft.Extensions.Hosting.AspireKafkaConsumerExtensions.AddKafkaConsumer<TKey, TValue>(this Microsoft.Extensions.Hosting.IHostApplicationBuilder! builder, string! connectionName) -> void
```
This is a lot of new APIs (we went from 4 to 24). Are all these needed?

g7ed6e replied:
We previously have 4 api with 2 optional parameters. To ensure backwards compatibility I dropped the default values
and exposed each combination of the previous API. we went from 4 to 4 * 4 APIs.

The new overloads replace `Action<ProducerBuilder<TKey, TValue>>` and `Action<ConsumerBuilder<TKey, TValue>>` parameters by `Action<IServiceProvider, ProducerBuilder<TKey, TValue>>` and `Action<IServiceProvider, ConsumerBuilder<TKey, TValue>>` which add 4 * 2 APIs.

At both side, consumer and producer, the APIs forward all the call to the same internal implementation.

g7ed6e committed:
Add unit test to cover ConfigureBuilder params

g7ed6e committed:
Update PublicAPI.Unshipped.txt

g7ed6e commented:
@mitchdenny / @eerhardt : PR is updated with latest changed from main :)

eerhardt committed:
Fix build error

Mark the public APIs with default values as removed to make PublicAPI analyzer happy.

eerhardt commented:
> @mitchdenny / @eerhardt : PR is updated with latest changed from main :)
> 
> edit: build is broken stating that some shipped public API no longer exist. I've followed https://github.com/dotnet/roslyn/blob/main/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md but it looks that the analyzer is failing to map the new methods to existing ones. @joperezr may i ask some guidance about how to solve this.

I resolved this build error by adding the `*REMOVED*` heading to the APIs. The analyzer considers this a change because the default values were dropped.

eerhardt commented on src/Components/Aspire.Confluent.Kafka/AspireKafkaConsumerExtensions.cs:
```diff
@@ -28,8 +28,110 @@ public static class AspireKafkaConsumerExtensions
     /// <remarks>Reads the configuration from "Aspire:Kafka:Consumer" section.</remarks>
-    public static void AddKafkaConsumer<TKey, TValue>(this IHostApplicationBuilder builder, string connectionName, Action<KafkaConsumerSettings>? configureSettings = null, Action<ConsumerBuilder<TKey, TValue>>? configureBuilder = null)
-        => AddKafkaConsumer(builder, DefaultConfigSectionName, configureSettings, configureBuilder, connectionName, serviceKey: null);
+    public static void AddKafkaConsumer<TKey, TValue>(this IHostApplicationBuilder builder, string connectionName, Action<KafkaConsumerSettings>? configureSettings, Action<ConsumerBuilder<TKey, TValue>>? configureBuilder)
```
1. Can we put the overloads in some sort of logical order? Potentially from simplest to most complex? And group all the `AddKafkaConsumer` overloads together and all the `AddKeyedKafkaConsumer` overloads together.

2. Can we use `/// <inheritdoc/>` so we don't have to duplicate all the XML comments so many times?

g7ed6e replied:
done!

eerhardt commented on tests/Aspire.Confluent.Kafka.Tests/ConsumerConfigurationTests.cs:
```diff
@@ -102,14 +102,96 @@ public void ConnectionNameWinsOverConfigSection(bool useKeyed)
         Assert.Equal(CommonHelpers.TestingEndpoint, config.BootstrapServers);
     }
 
+    [ConditionalTheory]
```
Why are these `ConditionalTheory`? I don't think they use SkipExceptions.

```suggestion
    [Theory]
```

g7ed6e replied:
done for all UT in that file. Others were not concerned

eerhardt commented on tests/Aspire.Confluent.Kafka.Tests/ProducerConfigurationTests.cs:
```diff
@@ -95,14 +95,90 @@ public void ConnectionNameWinsOverConfigSection(bool useKeyed)
+        Assert.Equal(CommonHelpers.TestingEndpoint, config.BootstrapServers);
+        return;
+
+        void ConfigureBuilder(IServiceProvider _, ProducerBuilder<string, string> __)
```
It would be good if the test actually used the IServiceProvider to ensure it works correctly.

g7ed6e replied:
done.

g7ed6e committed:
Use IServiceProvider in UT

g7ed6e committed:
Sort overloads and deduplicate xml docs

g7ed6e committed:
Use Theory instead of ConditionalTheory

g7ed6e commented:
> > @mitchdenny / @eerhardt : PR is updated with latest changed from main :)
> > edit: build is broken stating that some shipped public API no longer exist. I've followed https://github.com/dotnet/roslyn/blob/main/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md but it looks that the analyzer is failing to map the new methods to existing ones. @joperezr may i ask some guidance about how to solve this.
> 
> I resolved this build error by adding the `*REMOVED*` heading to the APIs. The analyzer considers this a change because the default values were dropped.

Thank you! i did not know that.

eerhardt commented on src/Components/Aspire.Confluent.Kafka/AspireKafkaProducerExtensions.cs:
```diff
@@ -28,8 +48,43 @@ public static class AspireKafkaProducerExtensions
+        AddKafkaProducerInternal<TKey, TValue>(builder, $"{DefaultConfigSectionName}:{name}", configureSettings, null, connectionName: name, serviceKey: name);
+    }
+
+    /// <inheritdoc cref="AddKafkaProducerInternal{TKey, TValue}(IHostApplicationBuilder, string, Action{KafkaProducerSettings}?, Action{IServiceProvider, ProducerBuilder{TKey, TValue}}?, string, string?)"/>
+    public static void AddKeyedKafkaProducer<TKey, TValue>(this IHostApplicationBuilder builder, string name, Action<ProducerBuilder<TKey, TValue>>? configureBuilder)
+    {
+        ArgumentException.ThrowIfNullOrEmpty(name);
+        AddKafkaProducerInternal<TKey, TValue>(builder, $"{DefaultConfigSectionName}:{name}", null, Wrap(configureBuilder), connectionName: name, serviceKey: name);
+    }
+
+    /// <inheritdoc cref="AddKafkaProducerInternal{TKey, TValue}(IHostApplicationBuilder, string, Action{KafkaProducerSettings}?, Action{IServiceProvider, ProducerBuilder{TKey, TValue}}?, string, string?)"/>
+    public static void AddKeyedKafkaProducer<TKey, TValue>(this IHostApplicationBuilder builder, string name, Action<IServiceProvider, ProducerBuilder<TKey, TValue>>? configureBuilder)
+    {
+        ArgumentException.ThrowIfNullOrEmpty(name);
+        AddKafkaProducerInternal<TKey, TValue>(builder, $"{DefaultConfigSectionName}:{name}", null, configureBuilder, connectionName: name, serviceKey: name);
+    }
+
+    /// <inheritdoc cref="AddKafkaProducerInternal{TKey, TValue}(IHostApplicationBuilder, string, Action{KafkaProducerSettings}?, Action{IServiceProvider, ProducerBuilder{TKey, TValue}}?, string, string?)"/>
```
You don't want to inhertdoc from `AddKafkaProducerInternal`. That method doesn't have any xml doc.

g7ed6e replied:
@eerhardt good catch. Sorry for that. this is now fixed.

g7ed6e committed:
Fix incorrect inheritdoc cref

g7ed6e commented:
> Looks like we took a dip in code coverage.

Hum, this PR replace methods with default parameters by methods with all combinations of those parameters. The code coverage was not really meaningful for some parameter combinations. 
I can write test to reach 90% code coverage if you think it makes sense.

g7ed6e committed:
Increase code coverage

g7ed6e commented:
@eerhardt @mitchdenny  i updated UT to cover all parameter combinations as it was before that PR.

mitchdenny commented:
/azp run

radical commented on tests/Aspire.Confluent.Kafka.Tests/ConsumerConfigurationTests.cs:
```diff
@@ -102,14 +102,111 @@ public void ConnectionNameWinsOverConfigSection(bool useKeyed)
     }
 
+    [Theory]
+    [InlineData(true, true, false, false)]
```
I think it would be easier to read if we added comments showing which overload each of these corresponds to.

g7ed6e replied:
Wouldn't that duplicate the content of the switch expression ?

radical replied:
Good point.

g7ed6e replied:
I think the way to get this more explicit is to drop the parameterization of the UT and explode combinations in dedicated UTs.. but that will duplicate the testing code.

radical commented on tests/Aspire.Confluent.Kafka.Tests/ConsumerConfigurationTests.cs:
```diff
@@ -102,14 +102,111 @@ public void ConnectionNameWinsOverConfigSection(bool useKeyed)
+        Action act =
+            (useKeyed, useConfigureSettings, useConfigureBuilder, useConfigureBuilderWithServiceProvider) switch
+            {
+                (true, true, false, false) => () => builder.AddKeyedKafkaConsumer<string, string>("messaging",
+                    configureSettings: ConfigureSettings),
+                (true, true, true, false) => () => builder.AddKeyedKafkaConsumer<string, string>("messaging",
+                    configureSettings: ConfigureSettings, configureBuilder: ConfigureBuilder),
+                (true, true, false, true) => () => builder.AddKeyedKafkaConsumer<string, string>("messaging",
+                    configureSettings: ConfigureSettings, configureBuilder: ConfigureBuilderWithServiceProvider),
+                (true, false, false, false) => () =>
+                    builder.AddKeyedKafkaConsumer<string, string>("messaging"),
+                (true, false, true, false) => () =>
+                    builder.AddKeyedKafkaConsumer<string, string>("messaging", configureBuilder: ConfigureBuilder),
+                (true, false, false, true) => () =>
+                    builder.AddKeyedKafkaConsumer<string, string>("messaging",
+                        configureBuilder: ConfigureBuilderWithServiceProvider),
+                (false, true, false, false) => () => builder.AddKafkaConsumer<string, string>("messaging",
+                    configureSettings: ConfigureSettings),
+                (false, true, true, false) => () => builder.AddKafkaConsumer<string, string>("messaging",
+                    configureSettings: ConfigureSettings, configureBuilder: ConfigureBuilder),
+                (false, true, false, true) => () => builder.AddKafkaConsumer<string, string>("messaging",
+                    configureSettings: ConfigureSettings, configureBuilder: ConfigureBuilderWithServiceProvider),
+                (false, false, false, false) => () =>
+                    builder.AddKafkaConsumer<string, string>("messaging"),
+                (false, false, true, false) => () =>
+                    builder.AddKafkaConsumer<string, string>("messaging", configureBuilder: ConfigureBuilder),
+                (false, false, false, true) => () =>
+                    builder.AddKafkaConsumer<string, string>("messaging",
+                        configureBuilder: ConfigureBuilderWithServiceProvider),
```
This is just a suggestion. What do you think of formatting it a little differently to help readability? like https://gist.github.com/radical/b168cca070fbe1fae392ae8d99480d34

This helps to see all the overloads being covered, and you can easily see that keyed/non-keyed both have been covered. And if one was missing it would be easy to spot.

g7ed6e replied:
Looks nice. I formatted both producer and consumer parameterized tests that way.

g7ed6e committed:
format parameterized tests to ease readability

mitchdenny commented:
/azp run

mitchdenny reviewed (approved):
LGTM. The overload explosion is unfortunate but lessons learned about API design I say ;)

g7ed6e commented:
> LGTM. The overload explosion is unfortunate but lessons learned about API design I say ;)

I agree.. I guess this is the drawback of defaulting multiple parameters!

eerhardt reviewed (approved):
LGTM. Thanks for the fix @g7ed6e.
